### ISSUE TITLE
Help users choose a good pipeline & policy configuration

### ISSUE THREAD

**Issue by tmbo** [RasaHQ/rasa#5454]
The goal is to give users an easy option to get started. The assumption is that most of our users do not (and should not) need to customize the pipeline or the configuration when they start building a bot. Customizing policy & pipeline configuration options is a feature used to finetune a bot later in the development stage.

# Option A

Add a new command
```bash
rasa config suggest
```

... which will ask users about their expectations (do they need certain things or not) and analyze their training data, suggesting users to use a certain configuration. Later with experience users can move to more suitable pipelines. 

* asking the user questions might allow as to pick a better pipeline (is this the case?)
* most users will not use it (simply because it is a separate command, e.g. might make sense to integrate into init script to increase that fraction)

# Option B

Automatically choose pipeline & polcies if, and only if, they are not provided as part of the configuration.

The default init project should be changed to
```yaml
language: en
```

As soon as training is started, we will pick a configuration based on the training data (e.g. adding a FormPoliciy, if needed). Option A can be implemented on top of this, reusing the code to generate the configuration.

* allows us to improve models "under the hood" when roling out updates by improving the automatic configuration
* reduces the friction to use additional features of Rasa Open Source (e.g. forms, response selector)

**Comment by akelad** [comment]
ok so i would say the automatic thing makes sense for Core policies for sure. But I'm not so sure about doing that for the NLU pipeline - that's a bit more complicated. E.g. we're not gonna run language detection on their data :P

**Comment by tmbo** [comment]
even though I'd say we prob should do that, in this proposal the language key is still there (right at the top).

**Comment by akelad** [comment]
then that involves people opening up that file and changing it anyways...

**Comment by tmbo** [comment]
I am not sure what you mean, yes of course, if they do not want to use the laguage that we use by default in rasa init, they need to go into that file and change the language configuration value. This proposal is not about the language parameter but about the pipeline and policies.

**Comment by akelad** [comment]
yeah but if we want to have `rasa train` select an appropriate pipeline, then they would need to specify the language they are building their bot in - that would surely make sense to ask in a prompt?

**Comment by tmbo** [comment]
I really don't get what you mean. Nothing will change about the language handling. When you run `rasa init` you'll end up with a project that contains a `config.yml`, in that config yaml there will be a key `language: en`. If you want a different language, you need to change that ofc. This is exactly the same as we have it right now: if you want a different langauge, go to the config and change it.

**Comment by akelad** [comment]
yeah i know, but if we're going to have `rasa train` automatically select a pipeline then people would probably forget to do that? so what i'm saying is for that we should have a prompt to enter the language somewhere. Since we're talking about asking the user a series of questions vs auto configuration

**Comment by JEM-Mosig** [comment]
We could also have `rasa init` ask the user a few things ("setup wizard"). E.g. "what language is your data in?".

**Comment by tmbo** [comment]
we can't really do the language selection, as we currently only have a starter project for `en` (the one you get with rasa init). if we would want to support that for multiple languages we would need to create these for more languages

**Comment by wochinge** [comment]
@tmbo can you please link your prototype?

**Comment by tmbo** [comment]
@wochinge here you go, there isn't a lot implemented yet. I tried different approaches and in my experience the best workflow was this:

1. if the configuration does not exist or if it does not contain `pipeline` / `policies` autoconfiguration is used during training. if these properties are available, training is done as it is now
2. once the automatic configuration is created, it should be dumped to the original file --> this makes it easier for the user to customize it later on. here is an example of how that could look like:
    * original user configuration:
    ```yaml
    language: en
    ```
    * configuration after user ran `train` and the automatic configuration ran:
    ```yaml
    language: en

    ## *°*°*
    ## Rasa will try to select a good configuration given your project and data.
    ## If you want to customize these machine learning parts, just uncomment the below
    ## configuration and change it as you wish - this will disable the
    ## automatic configuration.
    ## Here is the configuration Rasa selected for the most recent training run:
    #
    # pipeline:
    #  - name: SpacyNLP
    #  - name: SpacyTokenizer
    #  - name: SpacyFeaturizer
    #  - name: RegexFeaturizer
    #  - name: LexicalSyntacticFeaturizer
    #  - name: CountVectorsFeaturizer
    #  - name: CountVectorsFeaturizer
    #    analyzer: "char_wb"
    #    min_ngram: 1
    #    max_ngram: 4
    #  - name: DIETClassifier
    #    epochs: 100
    #  - name: EntitySynonymMapper
    #  - name: ResponseSelector
    #    epochs: 100
    #
    # policies:
    #  - name: MemoizationPolicy
    #  - name: TEDPolicy
    #    max_history: 5
    #    epochs: 100
    #  - name: MappingPolicy
    #  - name: FormPolicy
    ## °*°*°
    ```
    The generated configuration is added to the original configuration, but commented out. The `°*°*` 
    markers are there to allow us to overwrite this section, if the user runs train again and we need to do 
    another round of auto-generation.

Challenges:
* finding a good set of policies / pipelines. I think a first setup can just use the above list (all the time). as a next step we should think about how to tune parameters, e.g. `epochs` -> this is for the research team to figure out
* make sure if we use the above pipeline all the time, the components do not throw any errors if e.g. there is no form or there is no responses in the training data
* configuration should mostly be moved to the domain (e.g. parameters to fallbackpolicy) to avoid the need to specify the pipeline just to configure them.

**Comment by wochinge** [comment]
cc @alwx @ricwo @chkoss

**Comment by wochinge** [comment]
The `ConfigGenerator` might also be called from within the `TrainingDataImporter` since the training data importer has the whole picture of the other files

Suggested next steps:
1. Set up a short implementation proposal
  - addressing the challenges
  - thinking about further challenges
  - how are configurations moved / how is their deprecation handled
  - ...
2. Implement a first version (discussions for better suggestion rules should happen on the side)

**Comment by tmbo** [comment]
we also discussed multilanguage earlier and asking the user about the language as part of rasa init. I'd treat that separately, as it requires us to maintain multiple "init projects", the current initial data is only in available in `en`.

**Comment by tmbo** [comment]
@tabergma can we come up with a simple schema for the start where we pick decent values for `epochs` and `max_history`? (and any other parameters that we could easily improve)

I think we should start with something simple ala, 100 training samples -> 100 epochs, 1000 training samples -> 20 epochs or something like that.

**Comment by tabergma** [comment]
I think the default hyper-parameters we have now for the `DIETClassifier`, `ResponseSelector` and the `EmbeddingIntentClassifier` don't need to be touched. I think in most cases `epochs: 100` should work well. If the user has a lot of data (> 5000) then it might make sense to increase the number to 200. However a better way would be that we automatically stop training if the model is not improving anymore. Then we could just set the number of epochs to 200 or even higher and training would automatically stop earlier if the model "finished" training. But this is not yet implemented.

I don't have much experience with `max_history`, I guess maybe @Ghostvv can give some more insights here.

**Comment by Ghostvv** [comment]
automatic training stopping is tricky considering we have random negative sampling in the loss
regarding max_history, I think the default should be infinity (currently implemented by @evgeniiaraz for end-to-end, can be easily implemented in core right now)

**Comment by chkoss** [comment]
Closing this to keep the Production board tidy. All the individual steps will be tracked in their own issues -> #5820
